cpp/src/decisiontree/quantile/quantile.cuh

```diff
@@ -183,5 +183,69 @@ void preprocess_quantile(const T *data, const unsigned int *rowids,
   return;
 }
 
+template <typename T>
+__global__ void computeQuantilesSorted(T *quantiles, const int n_bins,
+                                       const T *sorted_data, const int length) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  double bin_width = static_cast<double>(length) / n_bins;
+  int index = int(round((tid + 1) * bin_width)) - 1;
+  // Old way of computing quantiles. Kept here for comparison.
+  // To be deleted eventually
+  // int index = (tid + 1) * floor(bin_width) - 1;
+  if (tid < n_bins) {
+    quantiles[tid] = sorted_data[index];
+  }
+
+  return;
+}
+
+template <typename T>
+void computeQuantiles(
+  T *quantiles, int n_bins, const T *data, int n_rows, int n_cols,
+  const std::shared_ptr<MLCommon::deviceAllocator> device_allocator,
+  cudaStream_t stream) {
+  // Determine temporary device storage requirements
+  MLCommon::device_buffer<char> *d_temp_storage = nullptr;
+  size_t temp_storage_bytes = 0;
+
+  MLCommon::device_buffer<T> *single_column_sorted;
+  single_column_sorted =
+    new MLCommon::device_buffer<T>(device_allocator, stream, n_rows);
+
```

Comment:
I guess, here too `std::unique_ptr` should be used. Also for `d_temp_storage`.

Reply:
yes please. Makes sense.